src/na/device/Device.cpp

```diff
@@ -0,0 +1,499 @@
+    if ((prop) == (prop_name)) {                                               \
+      if ((value) != nullptr) {                                                \
+        if ((size) < strlen(prop_value) + 1) {                                 \
+          return QDMI_ERROR_INVALIDARGUMENT;                                   \
+        }                                                                      \
+        strncpy(static_cast<char*>(value), prop_value, size);                  \
+        /* NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic) */  \
+        static_cast<char*>(value)[size - 1] = '\0';                            \
+      }                                                                        \
+      if ((size_ret) != nullptr) {                                             \
+        *size_ret = strlen(prop_value) + 1;                                    \
+      }                                                                        \
+      return QDMI_SUCCESS;                                                     \
+    }                                                                          \
+  }
+
+#define ADD_LIST_PROPERTY(prop_name, prop_type, prop_values, prop, size,       \
+                          value, size_ret)                                     \
+  {                                                                            \
+    if ((prop) == (prop_name)) {                                               \
+      if ((value) != nullptr) {                                                \
+        if ((size) < (prop_values).size() * sizeof(prop_type)) {               \
+          return QDMI_ERROR_INVALIDARGUMENT;                                   \
+        }                                                                      \
+        memcpy(static_cast<void*>(value),                                      \
+               static_cast<const void*>((prop_values).data()),                 \
+               (prop_values).size() * sizeof(prop_type));                      \
+      }                                                                        \
+      if ((size_ret) != nullptr) {                                             \
+        *size_ret = (prop_values).size() * sizeof(prop_type);                  \
+      }                                                                        \
+      return QDMI_SUCCESS;                                                     \
+    }                                                                          \
+  }
+// NOLINTEND(bugprone-macro-parentheses)
+
+namespace qdmi {
+Device::Device() {
+  // NOLINTNEXTLINE(cppcoreguidelines-prefer-member-initializer)
+  INITIALIZE_NAME(name);
+  // NOLINTNEXTLINE(cppcoreguidelines-prefer-member-initializer)
+  INITIALIZE_QUBITSNUM(qubitsNum);
+  INITIALIZE_OPERATIONS(operations);
+  INITIALIZE_SITES(sites);
+}
+Device::~Device() { sessions.clear(); }
```

Comment:
Is this even necessary? The default constructor will delete the sessions anyway, right?